### DISCUSSION TITLE
Scratch-like Visual Scripting

### DISCUSSION THREAD

**Discussion by lunabunn** [emilk/egui#305] [General]
I was wondering how feasible it would be to implement a scratch-like visual scripting system (online example: [Blockly](https://developers.google.com/blockly)) on top of egui. It seems pretty feasible.

1. Have a "block palette" widget that lets the user choose which block they want to insert, more or less a radio button functionality-wise
2. Have an actual editor widget. I'd probably want to leverage widgets like selection dropdowns and text input boxes within the editor though, so would have to think about it some more.
3. Show a drag-and-drop block overlay between 1 and 2. Only place blocks when the mouse is pressed on widget 1 and subsequently released on widget 2.

In my head the biggest blocker would be dragging-and-dropping blocks across widgets. The drag and drop would be a hundred percent visual. Is there a "nice" way to achieve this in egui only? Also, as I'm not super experienced with egui; what are yall's thoughts on number 2 above? How would yall implement it?

**Comment by emilk** [comment]
I don't think anyone has tried this yet, but it is an interesting use-case I'd like to support.

You keep track of the blocks and use [`ui.put`](https://docs.rs/egui/0.11.0/egui/struct.Ui.html#method.put) to place widgets where they belong. To place a whole `Ui` (e.g. a layout), you can use [`Ui::new`](https://docs.rs/egui/0.11.0/egui/struct.Ui.html#method.new).

There is a drag-and-drop demo at https://emilk.github.io/egui/ that may be a start.

**Comment by logica-academy** [comment]
I am also interested in this topic.
My kid's programming school in Japan is now using a native app made by a company that develops games.
This app is better than scratch.
However, this app is not a web app.
I would like to make a web app version of this app by ourselves.
For that purpose, I started learning Rust a month ago.
And I found egui.
I think it has a lot of potential.
But, I just need to learn a lot more.

https://user-images.githubusercontent.com/38776509/118223723-e1330280-b4bc-11eb-8b4b-4d618e21cbe6.mp4

**Comment by logica-academy** [comment]
I've been trying to create a code block using egui's window. 
I want to make it so that if we move one block closer to another block in the panel, the two blocks will stick together.
To do this, I need the coordinates of all the windows in the panel.
Is there any way to get that?

![block](https://user-images.githubusercontent.com/38776509/120929676-e6cee180-c724-11eb-9958-8d22985e1ca8.PNG)
